### PR TITLE
通过 Quick 和 Nimble 在 Swift 中进行测试驱动开发

### LINKED ISSUE #3265: 通过 Quick 和 Nimble 在 Swift 中进行测试驱动开发
* 原文链接：[Test Driven Development (TDD) in Swift with Quick and Nimble](https://www.appcoda.com/tdd-quick-nimble/)
  * Markdown文件：[文件地址](https://github.com/xitu/gold-miner/blob/master/TODO/tdd-quick-nimble.md)
  * PR 地址：#
  * 文章分类：**iOS**
  ----
  * 翻译时间：3 天
  * 校对时间：1 天
  * 翻译积分：3.5 分
  * 校对积分：1 分
  * [积分有什么用](https://github.com/xitu/gold-miner/wiki)

### PR DESCRIPTION
**Pull request by 94haox**
译文翻译完成，resolve #3265

### CODE REVIEW

**Comment by swants** [comment]
校对认领

**Comment by fanyijihua** [comment]
@swants 好的呢 :beer:

**Comment by leviding** [comment]
@94haox resolve 对应的应该是本 pull request 对应的 Issue 的标号，例如本 pull request 对应的是 #3265 ，这样才能连接上。

**Comment by atuooo** [comment]
@leviding 校对认领

**Comment by fanyijihua** [comment]
@atuooo 妥妥哒 :beers:

File: TODO/tdd-quick-nimble.md [hunk 1]
```diff
@@ -50,43 +50,47 @@ func calculateAreaOfSquare(w: Int, h: Int) -> Double {
...
-## Pre-requisites
+## 前期准备
 
-Here are some pre-requisites before we proceed to the field:
+在我们继续下去之前，有些前期准备：
 
-* Xcode 8.3.3 with Swift 3.1 installed
-* Some prior experience working with Swift and iOS development
+* Swift3 环境和 8.3.3 版本的 Xcode
+* 有 Swift 和 iOS 开发的经验。
```
Comment (swants) [inline TODO/tdd-quick-nimble.md]: 陈列点的末尾符号统一下，要么都有， 要么都没。

File: TODO/tdd-quick-nimble.md [hunk 2]
```diff
@@ -104,9 +108,9 @@ class MoviesTableViewController: UITableViewController {
 }
 ```
 
-### Movies
+### 电影数据
```
Comment (swants) [inline TODO/tdd-quick-nimble.md]: 电影数据 翻译成 电影库 是不是好些

File: TODO/tdd-quick-nimble.md [hunk 3]
```diff
@@ -148,39 +152,41 @@ class MoviesDataHelper {
...
+提醒一下，到目前为止，我们并没有在项目中做任何有关 TDD 的配置。现在，让我们开始学习这篇教程的主要内容 Quick 和 Nimble 吧！
 
 ## Quick & Nimble
 
-**Quick** is a test development framework designed for Swift and Objective-C which builds on top of XCTest. It provides a [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) to write a test very much similar to [RSpec](https://github.com/rspec/rspec).
+**Quick** 是一个建立在 XCTest 上，为 Swift 和 Objective-C 设计的测试框架. 
+它通过 [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) 去编写非常类似于 [RSpec](https://github.com/rspec/rspec) 的测试用例。
 
-**Nimble** is like a partner to **Quick** where **Nimble** provides the matchers as assertions. For more information about the framework, checkout this [link](https://github.com/Quick/Quick).
+**Nimble** 就像是 **Quick** 的合作伙伴，它提供了匹配器作为断言。关于它的更多信息，请查看[这儿](https://github.com/Quick/Quick)
```
Comment (swants) [inline TODO/tdd-quick-nimble.md]: 合作伙伴 换为 搭档 是不是更好些?  只作为建议

File: TODO/tdd-quick-nimble.md [hunk 4]
```diff
@@ -148,39 +152,41 @@ class MoviesDataHelper {
...
 
-### Installing Quick & Nimble using Carthage
+### 使用Carthage 安装 Quick & Nimble
```
Comment (swants) [inline TODO/tdd-quick-nimble.md]: Carthage 缺少前空格

File: TODO/tdd-quick-nimble.md [hunk 5]
```diff
@@ -148,39 +152,41 @@ class MoviesDataHelper {
...
 ```
 
-The above is the `CartFile.private` I used to install my dependencies. If you have no prior experience with Carthage, please check out this [link](https://github.com/Carthage/Carthage).
+上面就是 `CartFile.private` 中的内容，我通过它来安装依赖。如果你不熟悉 Carthage，先看看[它](https://github.com/Carthage/Carthage)吧.
 
-Place `CartFile.private` in your project folder and run `carthage update`. This will clone the dependencies and you should get both frameworks in your `Carthage -> Build -> iOS` folder. Then, add both frameworks into both test targets. You will also need to go to Build Phases, hit the plus sign at the top left, and choose “New Copy Files Phase”. Set destination as “Frameworks” and add both frameworks in.
+将 `CartFile.private` 拖入你的项目目录，然后终端运行 `carthage update`。这个命令会克隆依赖，成功后，你可以在 `Carthage -> Build -> iOS` 找到它们。接着，将两个框架都添加到测试工程。你需要到 Build Phases 点击左上方的加号，并且选择 “New Copy Files Phase”。将它设置为 “Frameworks”，并且将两个框架都添加进去。
 
-Way to go! You now have your test libraries all set up!
+现在所有的设置都搞定了！
```
Comment (swants) [inline TODO/tdd-quick-nimble.md]: Way to go! 可以翻译出来  
> way to go [用于表示快乐、赞同或兴奋]行，好！
> way to go 有个很地道的表达，表示一种恭喜，称赞别人很棒。

**Comment by leviding** [comment]
@swants 老哥，校对的时候不要 single comment 哈

**Comment by swants** [comment]
@leviding 过完年,脑子被狗吃了  >_<

File: TODO/tdd-quick-nimble.md [hunk 6]
```diff
@@ -221,15 +228,15 @@ class MyMoviesTests: QuickSpec {
...
 ```
 
-Notice that here we have a `@testable` import of `MyMovies`, this line basically identifies the project target we are testing, and then allows us to import classes from there. As we will be testing the view layer of the table view controller, we need to grab an instance from the storyboard.
+需要注意的是，我们有一个对 `MyMovies` 的 `@testable` 引用，这行代码的目的是标记着我们在测试哪个项目，并且允许我们引用那里的类。由于我们需要测试控制器的视图层，所以需要从 storyboard 抓取一个实例。
 
-The **describe** closure kicks start my first umbrella grouping of tests cases that will be written for `MoviesTableViewController`.
+**describe** 闭包开始了我的第一个为`MoviesTableViewController`而写的组合测试用例。
 
-The **beforeEach** closure runs before every example is executed in the **describe** closure. So you can read it as _beforeEach test in MoviesTableViewController is executed, run these first._
+**beforeEach** 闭包将在在 **describe** 闭包中所有例子执行之前运行。所以你可以在其中写一些需要在 `MoviesTableViewController` 执行时首先运行的测试。
```
Comment (swants) [inline TODO/tdd-quick-nimble.md]: **beforeEach** 闭包将在在   多了个字

File: TODO/tdd-quick-nimble.md [hunk 7]
```diff
@@ -221,15 +228,15 @@ class MyMoviesTests: QuickSpec {
 }
 ```
 
-Notice that here we have a `@testable` import of `MyMovies`, this line basically identifies the project target we are testing, and then allows us to import classes from there. As we will be testing the view layer of the table view controller, we need to grab an instance from the storyboard.
+需要注意的是，我们有一个对 `MyMovies` 的 `@testable` 引用，这行代码的目的是标记着我们在测试哪个项目，并且允许我们引用那里的类。由于我们需要测试控制器的视图层，所以需要从 storyboard 抓取一个实例。
 
-The **describe** closure kicks start my first umbrella grouping of tests cases that will be written for `MoviesTableViewController`.
+**describe** 闭包开始了我的第一个为`MoviesTableViewController`而写的组合测试用例。
```
[the remark this answers is not in the document]
Comment (94haox) [reply TODO/tdd-quick-nimble.md]: 长句很蛋疼==

File: TODO/tdd-quick-nimble.md [hunk 8]
```diff
@@ -5,40 +5,40 @@
 > * 译者：
 > * 校对者：
 
-Writing tests of any sort on a mobile application project has not been a very popular choice and in fact, most mobile project teams actually want to avoid writing tests as much as possible to speed up development.
+在移动开发领域，编写测试用例并不常见，事实上，大多数移动开发团队为了加快开发速度，选择性的放弃了它。
```
Comment (atuooo) [inline TODO/tdd-quick-nimble.md]: as much as possible：尽可能地避免编写测试用例

File: TODO/tdd-quick-nimble.md [hunk 9]
```diff
@@ -50,43 +50,47 @@ func calculateAreaOfSquare(w: Int, h: Int) -> Double {
 }
 ```
 
-Test 2 now PASSED! Hurray!
+测试 2 现在也通过了！哇哦！
 
-This can then go on and on till you have handled all edge cases and you should also refactor to make your code better and passed all the tests writen.
+这些操作可以继续下去，一直到你处理了所有的极限情况。接下来你就应该重构你的代码，在保证所有的测试用例都能通过的情况下，让它看起来更好。
```
Comment (atuooo) [inline TODO/tdd-quick-nimble.md]: 1、all edge cases: 所有边界情况

2、接下来你就应该重构你的代码，在保证所有的测试用例都能通过的情况下，让它看起来更好。
=>  同时，你也应该重构代码来让代码更漂亮以及通过所有测试用例

File: TODO/tdd-quick-nimble.md [hunk 10]
```diff
@@ -50,43 +50,47 @@ func calculateAreaOfSquare(w: Int, h: Int) -> Double {
...
 
-Based on what we have discussed so far, we realised that not only does TDD produces quality code, It allows us to handle edge cases early. Additionally, it also allows for pair-programming to happen between 2 developers effectively where one can write the test, and the other can write code to pass the test. You can find out more about TDD through [Dotariel’s Blog Post](https://medium.com/@dotariel/5-reasons-i-love-test-driven-development-fc257d9093e2#.7eejsiuwg).
+基于我们上面讨论的，我们意识到，TDD 不仅仅能让我们写出高质量的代码，它也能让我们更早的处理极限情况。另外它还能让两个开发者通过一个写测试用例，另一个写通过测试代码的方式来进行结对编程。你可以在 [Dotariel’s Blog Post](https://medium.com/@dotariel/5-reasons-i-love-test-driven-development-fc257d9093e2#.7eejsiuwg) 找到更多有关于 TDD 的信息。
```
Comment (atuooo) [inline TODO/tdd-quick-nimble.md]: 1、更早的处理极限情况 => 更早地处理边界情况
2、”另外它还能让。。。“ 注意断句
3、”另一个写通过测试代码的方式来进行结对编程“ 这里有点不通顺

File: TODO/tdd-quick-nimble.md [hunk 11]
```diff
@@ -50,43 +50,47 @@ func calculateAreaOfSquare(w: Int, h: Int) -> Double {
...
+## 你会在本教程中学到什么？
 
-By the end of this tutorial, this should be your takeaways:
+在教程的结尾，你可以获得以下的知识：
 
-* Have basic understanding of why **TDD is good**.
-* Have basic understanding of **how Quick & Nimble works**.
-* Know how to use **Quick & Nimble to write a UI Test**.
-* Know how to use **Quick & Nimble to write a Unit Test**.
+* 对 **TDD 很棒**，有一个基础的认知。
```
Comment (atuooo) [inline TODO/tdd-quick-nimble.md]: 对 **TDD 很棒** => 对于为什么 **TDD 很棒**

File: TODO/tdd-quick-nimble.md [hunk 12]
```diff
@@ -50,43 +50,47 @@ func calculateAreaOfSquare(w: Int, h: Int) -> Double {
...
-* Xcode 8.3.3 with Swift 3.1 installed
-* Some prior experience working with Swift and iOS development
+* Swift3 环境和 8.3.3 版本的 Xcode
+* 有 Swift 和 iOS 开发的经验
 
-## Setting up our project
+## 配置我们的项目
 
-Let’s assume that we are given a task to develop a simple movie app that can list movies. So lets fire up Xcode and create a new Single View Application called **MyMovies**. We will check `Unit Tests`, we will revisit this target once we setup our libraries and view controllers.
+假设我们要开发一个能够列出电影的简单的电影类 app 。 首先打开 Xcode 并创建一个叫做 **MyMovies** 的单视图应用。勾选上 ```Unit Tests```，一旦我们配置好库和视图控制器，我们将重新访问这个目标。
```
Comment (atuooo) [inline TODO/tdd-quick-nimble.md]: "假设我们要开发一个能够列出电影的简单的电影类 app" （有点啰嗦）
=> 假设我们要开发一个能够展示电影列表的 app

File: TODO/tdd-quick-nimble.md [hunk 13]
```diff
@@ -357,21 +365,21 @@ class MoviesTableViewController: UITableViewController {
...
-* We wrote our first test to check for movies count, it _fail_.
-* We implement our logic to load the movies and then it _passes_.
-* We wrote our second test to check for correct display, it _fail_.
-* We implement our logic for displaying and then it _passes_.
-* We then take a break from tests and _refactor_.
+* 我们为了检测电影数量，编写了第一个测试用例，测试 _未通过_
+* 接着我们实现了加载电影的逻辑，然后测试 _通过_
+* 为了检测是否显示了正确的数据，我们编写了第二个测试，测试 _未通过_
+* 接着我们实现了显示逻辑，然后测试 _通过_
+* 最后我们停止了测试，并且进行了 _重构_
```
Comment (atuooo) [inline TODO/tdd-quick-nimble.md]: 参考译文排版规则：[斜体文字使用加粗样式代替](https://github.com/xitu/gold-miner/wiki/%E8%AF%91%E6%96%87%E6%8E%92%E7%89%88%E8%A7%84%E5%88%99%E6%8C%87%E5%8C%97#%E6%96%9C%E4%BD%93%E6%96%87%E5%AD%97%E4%BD%BF%E7%94%A8%E5%8A%A0%E7%B2%97%E6%A0%B7%E5%BC%8F%E4%BB%A3%E6%9B%BF)

File: TODO/tdd-quick-nimble.md (hunk 9 above)
Comment (94haox) [reply TODO/tdd-quick-nimble.md]: you should also refactor to make your code better and passed all the tests writen.

让代码看起来更漂亮和通过所有测试用例，并不是同级的关系，首先要保证通过所有的测试用例，关于代码整洁是这之后的事情。

File: TODO/tdd-quick-nimble.md (hunk 8 above)
Comment (atuooo) [reply TODO/tdd-quick-nimble.md]: as much as possible 语义上要比 "选择性地" 程度深很多，而且我感觉实际开发的时候是这样的 😂，能不写就不写。

Comment (94haox) [reply TODO/tdd-quick-nimble.md]: 基本上是放弃了😂

File: TODO/tdd-quick-nimble.md (hunk 9 above)
Comment (atuooo) [reply TODO/tdd-quick-nimble.md]: 就原文的意思来看，重构和通过测试是一起进行的。

**Comment by leviding** [comment]
@94haox 已经 merge 啦~ 快快麻溜发布到[掘金](https://juejin.im)然后给我发下链接，方便及时添加积分哟。

掘金翻译计划有自己的知乎专栏，你也可以投稿哈，推荐使用一个[好用的插件](https://chrome.google.com/webstore/detail/minimalist-markdown-edito/pghodfjepegmciihfhdipmimghiakcjf?hl=zh-CN)。
专栏地址：https://zhuanlan.zhihu.com/juejinfanyi
